### PR TITLE
FIX ThresholdOptimizer: move validations to fit

### PR DESCRIPTION
**Pull request by adrinjalali**
This PR moves validations from `__init__` to `fit`, and adds an `estimator_` attribute to store the fitted estimator. Towards #222. Was a part of #282.

### CODE REVIEW

File: fairlearn/postprocessing/_threshold_optimizer.py [hunk 1]
```diff
@@ -69,25 +70,13 @@ class ThresholdOptimizer(ClassifierMixin, BaseEstimator):
...
-        if self._constraints not in _SUPPORTED_CONSTRAINTS:
-            raise ValueError(NOT_SUPPORTED_CONSTRAINTS_ERROR_MESSAGE)
-
-        self._grid_size = grid_size
-        self._flip = flip
-        self._plot = plot
-        self._post_processed_predictor_by_sensitive_feature = None
+        self.grid_size = grid_size
+        self.flip = flip
+        self.plot = plot
```
Comment (romanlutz) [inline fairlearn/postprocessing/_threshold_optimizer.py]: FYI plot is gone with my PR which I hope to complete today provided Richard/Miro manage to get a look

File: fairlearn/postprocessing/_threshold_optimizer.py [hunk 2]
```diff
@@ -114,25 +111,25 @@ def fit(self, X, y, *, sensitive_features, **kwargs):
         else:
             y = [int(y_val) for y_val in y]
 
-        if self._estimator:
+        if self.estimator:
             # train estimator on data first
-            self._validate_estimator()
-            self._estimator.fit(X, y, **kwargs)
-            self._unconstrained_predictor = self._estimator
+            self.estimator_ = self.estimator.fit(X, y, **kwargs)
```
Comment (romanlutz) [inline fairlearn/postprocessing/_threshold_optimizer.py]: This is mostly for my own understanding: doesn't this mean that `self.estimator` and `self.estimator_` are the same thing?

Comment (romanlutz) [reply fairlearn/postprocessing/_threshold_optimizer.py]: Oh I guess we did the same thing before as well... nvm

File: fairlearn/postprocessing/_threshold_optimizer.py (hunk 1 above)
Comment (romanlutz) [reply fairlearn/postprocessing/_threshold_optimizer.py]: ... unless this goes in first of course

Comment (romanlutz) [reply fairlearn/postprocessing/_threshold_optimizer.py]: Played around with this and managed to fix the flake8 issues. If the pipelines/tests pass I'll complete

File: fairlearn/postprocessing/_threshold_optimizer.py (hunk 2 above)
Comment (adrinjalali) [reply fairlearn/postprocessing/_threshold_optimizer.py]: yeah this is what was done before. It should probably be instead cloning before fitting, but didn't wanna include that in this PR.

**Comment by adrinjalali** [comment]
Thanks for the final touches @romanlutz

**Pull request merged**